### PR TITLE
Add integration tests for scontrol reboot functionality

### PR DESCRIPTION
**Pull request by jdeamicis**
### Description of changes
* Add integration tests for scontrol reboot functionality

### Tests
* Integration tests for scontrol reboot via test_runner

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integration-tests/configs/common/common.yaml [hunk 1]
```diff
@@ -465,6 +465,12 @@ schedulers:
         instances: {{ common.INSTANCES_DEFAULT_X86 }}
         oss: ["alinux2"]
         schedulers: ["slurm"]
+  test_slurm.py::test_scontrol_reboot:
+    dimensions:
+      - regions: ["us-east-1"]
```
Comment (gmarciani) [inline tests/integration-tests/configs/common/common.yaml]: Is it necessary to execute this test in us-east-1? If possible it would be better to distribute tests also in other regions.

File: tests/integration-tests/configs/common/common.yaml [hunk 2]
```diff
@@ -465,6 +465,12 @@ schedulers:
...
+        instances: {{ common.INSTANCES_DEFAULT_X86 }}
+        oss: ["ubuntu2004"]
```
Comment (gmarciani) [inline tests/integration-tests/configs/common/common.yaml:472]: Is there any specific reason why you choose ubuntu2004?

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 3]
```diff
@@ -377,6 +377,70 @@ def test_slurm_memory_based_scheduling(
...
+    test_datadir,
+    scheduler_commands_factory,
+):
+    cluster_config = pcluster_config_reader()
+    cluster = clusters_factory(cluster_config)
+    remote_command_executor = RemoteCommandExecutor(cluster)
+    slurm_commands = scheduler_commands_factory(remote_command_executor)
+
+    # Clear clustermgtd logs before starting the tests
+    remote_command_executor.run_remote_command("sudo truncate -s 0 /var/log/parallelcluster/clustermgtd")
```
Comment (gmarciani) [inline tests/integration-tests/tests/schedulers/test_slurm.py:394]: why is this necessary?

File: tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+Scheduling:
+  Scheduler: {{ scheduler }}
+  {{ scheduler_prefix }}Queues:
+    - Name: queue1
+      Networking:
+        SubnetIds:
+          - {{ private_subnet_id }}
+      ComputeResources:
+        - Name: t2medium
+          InstanceType: t2.medium
```
Comment (gmarciani) [inline tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml]: Why not using a t2.micro (4x cheaper) or a t2.small (2x cheaper) for this type of test?

File: tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+          MinCount: 2
+          MaxCount: 4
+SharedStorage:
```
Comment (gmarciani) [inline tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml]: Why do we need a shared storage for this type of test?

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 6]
```diff
@@ -1557,3 +1621,156 @@ def _test_memory_based_scheduling_enabled_true(
...
+
+    # Reboot nodes according to this logic:
+    # - 1 static idle node
+    # - 1 static idle node asap
+    # - 1 dynamic idle node
+    # - 1 dynamic idle node asap
+    static_nodes, dynamic_nodes = get_partition_nodes(nodes_in_queue)
+    assert_that(len(static_nodes)).is_equal_to(2)
+    assert_that(len(dynamic_nodes)).is_equal_to(2)
+    slurm_commands.reboot_compute_node(static_nodes[0], False)
```
Comment (gmarciani) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: In order to improve readability, may you please specify named parameter `asap=[True/False]`?

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 7]
```diff
@@ -1557,3 +1621,156 @@ def _test_memory_based_scheduling_enabled_true(
...
+    slurm_commands.reboot_compute_node(static_nodes[1], True)
+    slurm_commands.reboot_compute_node(dynamic_nodes[0], False)
+    slurm_commands.reboot_compute_node(dynamic_nodes[1], True)
+
+    # Check that nodes enter a reboot state
+    time.sleep(2)
```
Comment (gmarciani) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: 1. Is this wait time fair enough or we risk to check the state too early?
2. Isn't the wait time to check transition already implicit in `wait_for_compute_nodes_states()`
1. I see you always wait 2 seconds before checking state transitions. What about having such wait time defined in a variable so that we have a single place to control the wait time?

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 8]
```diff
@@ -1557,3 +1621,156 @@ def _test_memory_based_scheduling_enabled_true(
...
+    assert_compute_node_states(slurm_commands, nodes_in_queue, "reboot^")
+
+    # Wait that nodes come back after a while, without having triggered clustermgtd
```
Comment (gmarciani) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: It's not clear to me the relation between this comment and the actual check in line 1658. 
Why do we consider the absence of the log line `Found the following unhealthy static nodes` as a signal that clustermgtd has not been triggered?

File: tests/integration-tests/tests/schedulers/test_slurm.py [hunk 9]
```diff
@@ -1557,3 +1621,156 @@ def _test_memory_based_scheduling_enabled_true(
...
+    wait_for_compute_nodes_states(slurm_commands, nodes_in_queue, expected_states=["idle"])
+    assert_no_msg_in_logs(
+        remote_command_executor,
+        ["/var/log/parallelcluster/clustermgtd"],
+        ["Found the following unhealthy static nodes"],
+    )
+
+
+def _test_scontrol_reboot_alloc_nodes(
```
Comment (gmarciani) [inline tests/integration-tests/tests/schedulers/test_slurm.py]: The logic in this function seems very similar to the logic in `_test_scontrol_reboot_idle_nodes()`.
Do you think it could make sense to generalize i a single function?

File: tests/integration-tests/configs/common/common.yaml (hunk 2 above)
Comment (jdeamicis) [reply tests/integration-tests/configs/common/common.yaml:472]: No. I just wanted to put something different from `alinux2`. Or we can stick to `alinux2` if we prefer.

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 3 above)
Comment (jdeamicis) [reply tests/integration-tests/tests/schedulers/test_slurm.py:394]: Because as of now static nodes are always marked as unhealthy at the beginning of the life of the cluster, during the first iteration of clustermgtd. So we have to remove those messages if we want the test to be meaningful.

Alternatively we could build a function that checks the logs only from a certain point in time - which is out of scope for the time being, and also probably quite complicated to generalize for all possible log files we might want to analyze.

File: tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml (hunk 4 above)
Comment (jdeamicis) [reply tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml]: Addressing in the next commit.

File: tests/integration-tests/configs/common/common.yaml (hunk 1 above)
Comment (jdeamicis) [reply tests/integration-tests/configs/common/common.yaml]: Discussed separately: we are going to set `us-west-2`. Addressing in the next commit.

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 6 above)
Comment (jdeamicis) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: Agreed. Addressing in the next commit.

File: tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml (hunk 5 above)
Comment (jdeamicis) [reply tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml]: We really don't. I copied a fairly basic configuration from another slurm test and left it there. Addressing in the next commit.

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 7 above)
Comment (jdeamicis) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: 1. No, for such clusters 2 seconds should always be enough. The transition should be instantaneous, but I'm putting some buffer for safety.
2. We are doing an `assert_compute_node_states` before the `wait_for_compute_nodes_states`.
3. Addressing in the next commit.

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 8 above)
Comment (jdeamicis) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: Because if it had been triggered (i.e. if the node had been marked as unhealthy), you would see it in the logs with `Found the following unhealthy static nodes (x1) : ['nodename'(..."`, which comes from [here](https://github.com/aws/aws-parallelcluster-node/blob/9b9e1145fe212fb7b079821453e969b3adb5ae23/src/slurm_plugin/clustermgtd.py#L734).

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 9 above)
Comment (jdeamicis) [reply tests/integration-tests/tests/schedulers/test_slurm.py]: I was afraid generalizing it would have made the test workflow less readable, but indeed it came out fine. Addressed in the next commit.

File: tests/integration-tests/configs/common/common.yaml (hunk 2 above)
Comment (gmarciani) [reply tests/integration-tests/configs/common/common.yaml:472]: That's fine in both cases, just wanted to understand the reasoning behind the choice :-)

File: tests/integration-tests/tests/schedulers/test_slurm.py (hunk 3 above)
Comment (gmarciani) [reply tests/integration-tests/tests/schedulers/test_slurm.py:394]: Agree, let's stick with the current implementation.

File: tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml (hunk 5 above)
Comment (gmarciani) [reply tests/integration-tests/tests/schedulers/test_slurm/test_scontrol_reboot/pcluster.config.yaml]: Ok, so let's remove it if it's not necessary

**Review by gmarciani — APPROVED** [review]
